base/sparse/spqr.jl

```diff
@@ -21,133 +21,293 @@ const ORDERING_BESTAMD = Int32(9) # try COLAMD and AMD; pick best#
+
+# Examples
+```jldoctest
+julia> A = sparse([1,2,3,4], [1,1,2,2], ones(4))
+4×2 SparseMatrixCSC{Float64,Int64} with 4 stored entries:
+  [1, 1]  =  1.0
+  [2, 1]  =  1.0
+  [3, 2]  =  1.0
+  [4, 2]  =  1.0
+
+julia> qrfact(A)
+Base.SparseArrays.SPQR.QRSparse{Float64,Int64}(
+  [1, 1]  =  1.0
+  [4, 1]  =  0.414214
+  [2, 2]  =  1.0
+  [3, 2]  =  0.414214, [1.70711, 1.70711],
+  [1, 1]  =  -1.41421
+  [2, 2]  =  -1.41421, [1, 2], [1, 4, 2, 3])
+```
+"""
+Base.LinAlg.qrfact(A::SparseMatrixCSC; tol = _default_tol(A)) = qrfact(A, Val{true}, tol = tol)
+
+Base.LinAlg.qr(A::SparseMatrixCSC; tol = _default_tol(A)) = qr(A, Val{true}, tol = tol)
+
+function Base.A_mul_B!(Q::QRSparseQ, A::StridedVecOrMat)
+    if size(A, 1) != size(Q, 1)
+        throw(DimensionMismatch("size(Q) = $(size(Q)) but size(A) = $(size(A))"))
     end
-    f = Factorization(size(A)..., ccall((:SuiteSparseQR_C_factorize, :libspqr), Ptr{C_Factorization{Tv}},
-        (Cint, Cdouble, Ptr{Sparse{Tv}}, Ptr{Void}),
-            ordering, tol, get(A.p), common()))
-    finalizer(f, free!)
-    f
+    for l in size(Q.factors, 2):-1:1
+        τl = -Q.τ[l]
+        h = view(Q.factors, :, l)
+        for j in 1:size(A, 2)
+            a = view(A, :, j)
+            LinAlg.axpy!(τl*dot(h, a), h, a)
+        end
+    end
+    return A
 end
 
-function solve(system::Integer, QR::Factorization{Tv}, B::Dense{Tv}) where Tv<:VTypes
-    m, n = size(QR)
-    mB = size(B, 1)
-    if (system == RX_EQUALS_B || system == RETX_EQUALS_B) && m != mB
-        throw(DimensionMismatch("number of rows in factorized matrix must equal number of rows in right hand side"))
-    elseif (system == RTX_EQUALS_ETB || system == RTX_EQUALS_B) && n != mB
-        throw(DimensionMismatch("number of columns in factorized matrix must equal number of rows in right hand side"))
+function Base.A_mul_B!(A::StridedMatrix, Q::QRSparseQ)
+    if size(A, 2) != size(Q, 1)
+        throw(DimensionMismatch("size(Q) = $(size(Q)) but size(A) = $(size(A))"))
+    end
+    tmp = similar(A, size(A, 1))
+    for l in 1:size(Q.factors, 2)
+        τl = -Q.τ[l]
+        h = view(Q.factors, :, l)
+        A_mul_B!(tmp, A, h)
+        LinAlg.lowrankupdate!(A, tmp, h, τl)
     end
-    d = Dense(ccall((:SuiteSparseQR_C_solve, :libspqr), Ptr{C_Dense{Tv}},
-        (Cint, Ptr{C_Factorization{Tv}}, Ptr{C_Dense{Tv}}, Ptr{Void}),
-            system, get(QR.p), get(B.p), common()))
-    finalizer(d, free!)
-    d
+    return A
 end
 
-function qmult(method::Integer, QR::Factorization{Tv}, X::Dense{Tv}) where Tv<:VTypes
-    mQR, nQR = size(QR)
-    mX, nX = size(X)
-    if (method == QTX || method == QX) && mQR != mX
-        throw(DimensionMismatch("Q matrix size $mQR and dense matrix has $mX rows"))
-    elseif (method == XQT || method == XQ) && mQR != nX
-        throw(DimensionMismatch("Q matrix size $mQR and dense matrix has $nX columns"))
+function Base.Ac_mul_B!(Q::QRSparseQ, A::StridedVecOrMat)
```

Comment:
I did a few comparisons and the Julia implementations are faster. Probably because they avoid the conversion to `CHOLMOD.Sparse`.